File: algorithms/438. Find All Anagrams in a String/solution.py
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        counter = [0] * 26
        for c in p:
            counter[ord(c) - ord('a')] += 1
        window = [0] * 26
        n, m = len(s), len(p)
        result = []
        for i in xrange(n):
            window[ord(s[i]) - ord('a')] += 1
            if i >= m:
                window[ord(s[i - m]) - ord('a')] -= 1
            if counter == window:
                result.append(i - m + 1)
        return result

# 36 / 36 test cases passed.
# Status: Accepted
# Runtime: 155 ms

File: algorithms/438. Find All Anagrams in a String/solution.cc
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> result;
        vector<int> counter(26);
        for ( char c : p ) {
            ++counter[c - 'a'];
        }
        vector<int> window(26);
        int n = s.size(), m = p.size();
        for ( int i = 0; i < n; ++i ) {
            ++window[s[i] - 'a'];
            if ( i >= m ) {
                --window[s[i - m] - 'a'];
            }
            if ( window == counter ) {
                result.push_back(i - m + 1);
            }
        }
        return result;
    }
};

// 36 / 36 test cases passed.
// Status: Accepted
// Runtime: 33 ms
// beats 87.52%
